### ISSUE TITLE
[65816] 16-bit register mode support

### ISSUE THREAD

**Issue by asiekierka** [llvm-mos/llvm-mos#321]
Part of https://github.com/llvm-mos/llvm-mos/issues/32 .

The main challenge here is being clever about emitting REP/SEP instructions to switch, and deciding which register widths to support and how - while the 8-bit registers A and B are effectively subsets of their 16-bit counterpart, they are not as cheap to access as on most LLVM-supported architectures; in addition, the 8-bit registers X and Y cannot be treated as subsets of their 16-bit counterparts, as their high portions are reset when the 16-bit index register mode is disabled.

Note that a target which provides the 65816's 16-bit register modes, but not the 24-bit address space, is already present in llvm-mos-sdk as "rpc8e" (RedPower 2's RPC8/e "fantasy" computer), and may as such prove useful for testing.

**Comment by asiekierka** [comment]
Some nomenclature, first. The 65816 models the 16-bit accumulator as three distinct registers:

* `A` is the 8-bit accumulator register as used on the 6502,
* `C` is the 16-bit variant of the accumulator register, taking place of `A` when the 16-bit accumulator mode is enabled.
* `B` is the high portion of `C`; the high 8 bits. Its value (for the most part) can only be accessed in 8-bit mode by using `XBA`, which swaps `A` with `B`.

The 16-bit accumulator mode is often referred to as `a16`, with the 8-bit mode referred to as `a8`. The index register modes are referred to as `xy16`/`xy8` or `i16`/`i8` - I shall stick to the former.

Next, research. The ports of LCC (from 1994) and TCC (somewhat more recent) to generate code for the 65816 use `xy16` only, while toggling between `a16` (default) and `a8` in the only place where it matters significantly - for loading and storing variables. Modeling `A` is significantly easier, because:

* When switching between 16-bit and 8-bit accumulator modes, `B` is not destroyed; `XY` have their high bits 8-15 forced to `$00`.
* When switching between 16-bit and 8-bit accumulator modes, only `A` is affected; `XY` can only be affected as a pair.

A cursory analysis of one recent, major SNES codebase - [Nova the Squirrel 2](https://github.com/NovaSquirrel/NovaTheSquirrel2) shows that moving between `a8` and `a16` is much more common than toggling `xy8` and `xy16`, even in ASM. Based on further discussions with ASM developers, `xy8` is typically used for relatively niche optimizations, such as quick writes of consecutive 8-bit values, 256-byte table wraparound, or simply smaller code size/faster execution - all opcodes which access memory are 1 cycle slower and 1 byte larger if they need to access 16-bit memory, including loading/storing 16-bit immediates.

As such, it appears reasonable to split "16-bit register mode support" further into two stages:

1. Basic `xy16` and `a8`/`a16` support. For instructions and register allocation, model `X`, `Y`, `C` as 16-bit and `A` as 8-bit; at a late stage of code generation, emit REP and SEP opcodes based on the registers used by emitted instructions.
2. `xy8`/`xy16` support - this may well be a pipe dream with our current resources.

**Comment by asiekierka** [comment]
With regards to prior art, https://github.com/jackoalan/llvm-mos/commit/ec070a70ba8d8b3d3a9da24b4216435f9575f6bb#diff-1b1d72e821268f4e111b07ac7c40859bd1bd4d1a4279c189a7be4280f29583f3 provides a proof-of-concept implementation of such an "emit REP and SEP opcodes" pass.

**Comment by gingerbeardman** [comment]
I had a go at the accumulator half of this — 16-bit `A` only, leaving `X`/`Y` 8-bit since their high bytes reset anyway. It works, but the more useful result turned out to be a profitability heuristic, because "16-bit mode = faster" was a bit of a trap once I ran it on real code.

Branch, pinned to `c798c314`:
<https://github.com/gingerbeardman/llvm-mos/tree/mos-native-16bit-accumulator>
Write-up and how to reproduce:
<https://gist.github.com/gingerbeardman/03d232f637c02a30b16c42a302cf782c>

## How it works

It's a 65816-only `+native-16bit-accumulator` feature, off by default. No register-bank changes (operands already map by size) and no new instructions — `REP`/`SEP` and the `*_ZeroPage` forms already exist, and in `M=0` the same ZeroPage opcode is already 16-bit. i16 constants get materialised into an `Imag16` pair so the reg-reg form can be used.

The mode state was the part I expected to be fiddly, but it stays correct almost for free: each op expands (post-RA) to its own self-contained `REP #$20 … SEP #$20`, so `M=0` never leaves a single instruction and there's no cross-block / call / interrupt mode tracking to get wrong. A late peephole then strips the `STA/SEP/REP/LDA` between adjacent same-slot regions (a store-reload around a toggle that cancels out) and threads the value through `A`, so a whole chain collapses to one `REP` at the top and one `SEP` at the bottom.

## The good case

On a long i16 chain (synthetic kernel, 65535 calls on hardware) it's exactly what you'd hope for: 26% less time and 55% less code, ~1.35x. I'd stress the code-size half — on a ROM-bank-constrained target that isn't a footnote, it's more code fitting per bank and fewer cross-bank trampolines.

## Where it falls over

Then I turned it on across a whole real game (a LÖVE prototype I'm porting to SNES, transpiling Lua to C), same bitcode, only `-mattr` differing. Hot `.text` went *up* 17%, and the `-O2` LTO build overflowed its ROM bank by about 4 KB. Most of that game's i16 work is isolated single ops — plus the i16 adds the legalizer itself emits for pointer arithmetic — and each one pays a full `REP/SEP` with nothing adjacent to merge into. So the kernel number doesn't transfer; blanket-on it doesn't just run slower, it doesn't fit.

## The heuristic

So the real work was deciding, per op, whether 16-bit is worth it — and doing it before instruction selection. A pre-legalizer pass walks each i16 op's single-use data-flow chain and only keeps it native if the chain is at least `-mos-native16-min-chain` long (default 3, roughly the `REP/SEP` break-even). Those ops get rewritten to new target-generic opcodes (`G_NATIVE16_*`); everything else stays a plain i16 op and the legalizer narrows it to the normal register-resident byte-split — including the pointer-arithmetic adds it generates itself. It has to happen before selection: post-RA, even a one-op native region is already smaller than a *memory* byte-split, so there's no getting back to the cheap register-resident version after the fact.

With that in, the same game core comes out byte-identical to feature-off (the 17% and the overflow are gone — it has no chains ≥3 to begin with), and the chain kernel still compiles native and keeps the 1.35x. Forcing the threshold to 1 brings the overflow back, which is a reasonable sanity check that the gate is what's doing the work.

So for this particular game 16-bit-A buys nothing, but it can't hurt it either, and a chain-heavy program would get both the speed and the size win. That's the result I was actually after: safe to leave switched on, rather than a sharp edge.

## A few notes if anyone picks this up

- Making `G_NATIVE16_*` target-generic opcodes (like `G_SBC`) means they aren't `isPreISelGenericOpcode`, so the legalizer and its verifier leave them alone and the selector lowers them in C++. That also dodges the trap I hit the first time, where making plain i16 *legal* broke `G_UNMERGE_VALUES s64→s16` in a soft-float libcall.

- The 16-bit-immediate issue from #454 is real: in `M=0` the assembler still picks an 8-bit immediate for a small literal, so you'd have to force the 16-bit opcode (test `0x00ff`, not just values > `0xff`). I sidestepped it by materialising constants into a pair.

- The `CLC`/`SEC` inside a region need an explicit implicit-def of carry, otherwise the post-expansion verifier complains the `ADC`/`SBC` carry use is undefined — the merge peephole runs after that check.

- `X`/`Y` 16-bit, far pointers and the 24-bit space (#320) are all out of scope here; this is only the `M` flag.

Happy to turn it into a PR if there's appetite for landing the flag + accumulator slice plus the heuristic. The thing I mainly took away: the `REP`/`SEP` mechanics are the easy bit, deciding *when* to use them is the actual problem.
